pkg/cloud/services/networking/router.go

```diff
@@ -185,41 +186,41 @@ func setRouterExternalIPs(client *gophercloud.ServiceClient, openStackCluster *i
 	return nil
 }
 
-func (s *Service) DeleteRouter(openStackCluster *infrav1.OpenStackCluster, network *infrav1.Network) error {
-	if network.Router == nil || network.Router.ID == "" {
-		s.logger.V(4).Info("No need to delete router since no router exists.")
-		return nil
-	}
-	exists, err := s.existsRouter(network.Router.ID)
+func (s *Service) DeleteRouter(openStackCluster *infrav1.OpenStackCluster, clusterName string) error {
+	routerName := getRouterName(clusterName)
+	router, err := s.getRouterByName(routerName)
 	if err != nil {
 		return err
 	}
-	if !exists {
-		s.logger.Info("Skipping router deletion because router doesn't exist", "router", network.Router.ID)
-		return nil
+
+	subnetName := getSubnetName(clusterName)
+	subnet, err := s.getSubnetByName(subnetName)
```

Comment:
can we also return a pointer in getSubnetByName (please also check similar funcs in other places)